### PR TITLE
Multi Mode Support

### PR DESCRIPTION
**Pull request by ssayer**
Hey @whomwah,

Here's my first pass at addressing #23. It's not done, but I thought this was a good point to get feedback:

* Calculating the minimum version for a multi mode QR code means I have to iteratively calculate the length of the payload on the fly based on the actual input.
* Users must declare mode: multi and pass in an array of segments (see test for example input). I did this to ensure backwards compatibility. I had to move some code from the constructor into methods because I needed to reuse it for either size calculation or writing the QR code.
* Added support for a "max_size" option, in order to limit the size to a version other than 40. This isn't strictly related to multi mode support, but it's useful due to the need for calculation.
* I tried to keep my code separated where possible (e.g. `MultiUtil`, `MultiTest`), but I can move these into existing `qr_code`, `qr_util` etc. if you prefer.

Let me know what you think! I'll be working on documentation and cleanup in the meantime.

### CODE REVIEW

File: lib/rqrcode_core/qrcode/multi_util.rb
Comment (ssayer) [inline lib/rqrcode_core/qrcode/multi_util.rb:31]: After further investigation, this code does seem to be overestimating the amount of space needed for content. In testing known valid codes, it's now throwing an error saying the data is too long.

**Comment by whomwah** [comment]
Hey, thanks for this @ssayer. I'll try and get to look at this at my first opportunity.

**Comment by ssayer** [comment]
@whomwah in the meantime, can you approve me to run the workflows so I can see if all the tests/builds pass?

File: lib/rqrcode_core/qrcode/multi_util.rb
Comment (whomwah) [inline lib/rqrcode_core/qrcode/multi_util.rb:56]: (not blocking) maybe more Rubyish

```suggestion
((length / 2) * 11).tap do |size|
  size += 6 if length.odd?
end
```

File: test/rqrcode_core/multi_test.rb
Comment (whomwah) [inline test/rqrcode_core/multi_test.rb:14]: I'd never seen `flunk` before!

**Review by whomwah — COMMENTED** [review]
This looks great! Some of it is over my head to be honest, but the fact it's opt in and is tested is fine with me.

**Comment by whomwah** [comment]
I seem to have to  approve to run the CI every time. I wonder if it only remembers once you are a contributor?

**Comment by ssayer** [comment]
Thanks for the feedback, I'll get started on these items. I would like to add named params in a couple of places. I didn't see it used so I wasn't sure if that was a backwards compatibility thing, but it looks like we're targeting 2.5+ right?

**Comment by whomwah** [comment]
> I would like to add named params in a couple of places. I didn't see it used so I wasn't sure if that was a backwards compatibility thing, but it looks like we're targeting 2.5+ right?

Yes, `spec.required_ruby_version = ">= 2.3"`. It's purely a legacy thing where I've just not had the time to make the change.

**Comment by ssayer** [comment]
One thing I was considering is adding a `QRSegment` class to use instead of the hashes in the Array. The original constructor style w/ string could be mapped to use this internally in the constructor. This would allow the constructor to be used in different ways:

QRCode.new('foo', mode: 'byte_8bit')
QRCode.new(QRSegment.new(mode: 'byte_8bit', data: 'foo'))
QRCode.new([QRSegment.new(mode: 'byte_8bit', data: 'foo'), QRSegment.new(mode: 'alphanumeric', data: 'bar1')])

It's a little more verbose, but I think a bit cleaner and could consolidate some of the code related to handling modes.

File: lib/rqrcode_core/qrcode/multi_util.rb
Comment (ssayer) [reply lib/rqrcode_core/qrcode/multi_util.rb:56]: This doesn't seem to work for me. It looks like the tap method returns the original `self` no matter what you do in the block

Comment (ssayer) [reply lib/rqrcode_core/qrcode/multi_util.rb:59]: I did a little more digging on this. In order to update the original value, you have to use a method that modifies `self`. Integer/numbers in Ruby are immutable so you can't tap them and make any changes. I do think there is a more rubyish way to do this though, so I'll take another look at it today.

Comment (whomwah) [reply lib/rqrcode_core/qrcode/multi_util.rb:59]: You're right Samuel. I shouldn't have assumed. I guess I've only used before to alter my own objects or with strings! Like I say, it's not blocking but just doesn't look as "Ruby" as the rest.

Comment (whomwah) [reply lib/rqrcode_core/qrcode/multi_util.rb:59]: and this kind of thing may just make it more unreadable.
```
size = (length / 2) * 11
length.odd? ? size += 6 : size
```
Sometimes the simplest way is the best way I guess.

Comment (ssayer) [reply lib/rqrcode_core/qrcode/multi_util.rb]: I did refactor this to remove some redundant code. I'm not sure if it's more readable though...

**Comment by ssayer** [comment]
@whomwah ^ Can you let me know what you think of the QRSegment idea? Otherwise if you'd prefer to keep the code how it is, I'll mark this as ready.

**Comment by whomwah** [comment]
@ssayer I like the idea of the `QRSegment` class 👍

**Comment by ssayer** [comment]
Once this is merged, we should probably add an equivalent class in RQRCode for QRSegment like you do with QRCode so we're not mixing RQRCode and RQRCodeCore when using segments.

**Comment by whomwah** [comment]
Thanks. I should have time to have a good look at all these changes tomorrow at last. I'll also have a think about the RQRCode interface too. I'd really want to that interface to remain as simple to use as possible.

The RQRCode gem is a conversation on its own 😀

File: lib/rqrcode_core/qrcode/qr_code.rb
Comment (whomwah) [inline lib/rqrcode_core/qrcode/qr_code.rb:202]: I think on reflection that we should instantiate the `QRSegment`'s in the initializer. From a user point of view I think:

```
multi_qr = RQRCodeCore::QRCode.new([
  {data: 'foo', mode: :byte_8bit}, 
  {data: 'bar1', mode: :alphanumeric}
])
```

is less to write and care about than

```
multi_qr = RQRCodeCore::QRCode.new([
  RQRCodeCore::QRSegment.new(data: 'foo', mode: :byte_8bit), 
  RQRCodeCore::QRSegment.new(data: 'bar1', mode: :alphanumeric)
])
```

and from the `RQRCode` point of view (whilst multi mode is quite niche) I think not having to add any extra code in that gem is a good thing, especially as I'd like to move eventually to having the various renderers be light weight and handle a single rendering `rqrcode_png`, `rqrcode_svg` etc. I get a lot of requests for different renderings of a QR Code and at the moment I am pushing back as I don't want to turn `RQRCode` into an unmanageable monolith.

**Review by whomwah — CHANGES_REQUESTED** [review]
A few comments now I've had time to sit a look things over. Nothing too controversial I hope. Thanks again for the work you have done so far.

File: lib/rqrcode_core/qrcode/qr_code.rb
Comment (ssayer) [reply lib/rqrcode_core/qrcode/qr_code.rb:505]: I used the same table you referenced in the comments to create the `QRMAXBITS` list. Because multi-mode support can have any number of encodings, I count the total number of bits between all the segments and compare it to the maximum number of bits supported by the level/version. If there is a way to reuse `QRMAXDIGITS` for an arbitrary number of encoded segments, then I can remove it, but I couldn't figure that out. Theoretically you could use `QRMAXBITS` for everything, but I didn't want to touch the existing code.

File: lib/rqrcode_core/qrcode/qr_numeric.rb
Comment (ssayer) [reply lib/rqrcode_core/qrcode/qr_numeric.rb:30]: I made this change so I could use `QRNumeric::NUMBER_LENGTH` in `QRSegment`, but now that you point it out `private` doesn't actually protected the constant, just the methods. I can change it back.

File: lib/rqrcode_core/qrcode/qr_segment.rb
Comment (ssayer) [reply lib/rqrcode_core/qrcode/qr_segment.rb:20]: Yes, this is more Rubyish. I'll replace the existing code.

File: test/rqrcode_core/qr_segment_test.rb
Comment (ssayer) [reply test/rqrcode_core/qr_segment_test.rb:2]: Yes we should. I added this to `QRSegment` and `QRMulti`, looks like it's used in all the other classes in lib/

File: test/rqrcode_core/qr_segment_test.rb
Comment (ssayer) [reply test/rqrcode_core/qr_segment_test.rb:18]: Oops that was just to help me debug, I'll uncomment

File: lib/rqrcode_core/qrcode/qr_code.rb
Comment (ssayer) [reply lib/rqrcode_core/qrcode/qr_code.rb:202]: I think that's a good idea, I think it gives us the best of best worlds. I'll make the changes to do it this way.

File: lib/rqrcode_core/qrcode/qr_code.rb
Comment (whomwah) [reply lib/rqrcode_core/qrcode/qr_code.rb:504]: Thanks for the explanation. Any change that will also improve the existing code base would be welcome. The test suite in this lib and RQRCode is quite comprehensive so any problems should be easy to spot.

Comment (ssayer) [reply lib/rqrcode_core/qrcode/qr_code.rb:504]: OK, I can remove `QRMAXDIGITS` and just use the `QRMAXBITS` calculation all around if you want.

Comment (whomwah) [reply lib/rqrcode_core/qrcode/qr_code.rb:504]: Thanks. Maybe the complete removal should be left to another follow up PR? 

I'm away with my family for a week but I'll look at all this as soon as I am back.

Comment (ssayer) [reply lib/rqrcode_core/qrcode/qr_code.rb:412]: Enjoy your time off, this should be ready when you get back.

File: lib/rqrcode_core/qrcode/qr_code.rb
Comment (whomwah) [inline lib/rqrcode_core/qrcode/qr_code.rb:410]: Don't think we need these comments now?

**Review by whomwah — CHANGES_REQUESTED** [review]
Really close now. Some very tiny comments.

**Comment by ssayer** [comment]
Sounds good, I'll know these out later today.

File: lib/rqrcode_core/qrcode/qr_code.rb
Comment (ssayer) [reply lib/rqrcode_core/qrcode/qr_code.rb:53]: Should I change this to mention QRMAXBITS or remove it? QRMAXBITS still only goes up to version 40 like QRMAXDIGITS

Comment (whomwah) [reply lib/rqrcode_core/qrcode/qr_code.rb:53]: Good idea, let's switch to mention QRMAXBITS 👍

**Comment by ssayer** [comment]
I think that covers all the comments. I'll be available for any further changes until Thursday, when I leave for vacation until the 23rd.

**Comment by whomwah** [comment]
Just seen your comment. I will look over this and do a test with `RQRcode` and then I think it will be ready to merge. Thank you for this piece of work and your patience getting it added. I'll wait until you are back from your vacation before I release a new version of the Gem.

**Review by whomwah — DISMISSED** [review]
This all looks great! 👍

**Review by whomwah — APPROVED** [review]

**Comment by ssayer** [comment]
That's great! I'm back from vacation as of today, so let me know if you need anything.

**Comment by whomwah** [comment]
I'm due to be working on the project this Thu so I will likely do a new release then 👍 

PS I know this is not related but I'd be interested in your opinion if you have any spare time: https://github.com/whomwah/rqrcode_core/pull/28

**Comment by ssayer** [comment]
Like you, I'm not an expert in bitwise operators but it would make sense that using them would be more efficient than doing math with a bunch of integer literals in Ruby (though I am surprised it makes that much difference in memory).

In regards to your comment about hardcoding to 32bits: I suppose you could try hardcoding the result of `num.size  * 8`  into a constant in the library and just use that instead of running num.size every single time. I'm a little surprised using either 32 or 64b it works, but like I said I'm not an expert.

If you want to hold off a little on a release till both these are in, I'm not in a huge hurry. For the time being I can just point our Gemfile at GitHub.

**Pull request merged**
